### PR TITLE
feat: step functions event source, `when` refining, `when` and `map` narrowing 

### LINKED ISSUE #71: Support `when` narrowing on event bridge.
The `.when` method in event bridge does not narrow the type of the `map` or `pipe` methods.

```ts
interface Detail {
   optional?: string;
}

bus
   .when(event => event.detail.optional !== undefined)
   .map<string>(event => event.detail.optional) // compiler error because optional could be undefined
```

### LINKED ISSUE #99: Event Bridge - Chained Rules
```ts
const rule = new EventBus().when(event => event.source == "lambda");
const highRule = rule.when(event => event.detail.level === "high");
const lowRule = rule.when(event => event.detail.level === "low");
```

### Why?

In general this is a nice to have, however, this addition will make Event Sourced (#44) much more useful by further filtering rules created by services.

```
new StepFunction().onError(stack, "id").when(event => event.detail.error.startWith("Fault"));
```

> Note: this will not work for imported rules, only rules created using `new EventBusRule()` or `IEventBus.when()`. (until we support decompiling EB Patterns to Functionless Patterns.)

* Need to maintain the unresolved functionless event pattern
* Need to apply an AND between each of the chained `when` statements
* Need to figure out what to do with the stack and id parameter. It may not be used as the rule created may not be needed.
   * Consider only creating the rule when a target is added.

### PR DESCRIPTION
**Pull request by thantos**
Closes #71 Closes #99
Part Of #44 

Depends on PR #100 

* [x] Support Narrowing events in `when` and `map`
* [x] Support refining event bridge events
* [x] Support `on*` events for Step Functions
   * [x] Success
   * [x] Failure
   * [x] Status Changed
   * [x] Started
   * [x] Timeout
   * [x] Aborted

## Narrowing

`when` and `map` now correctly narrow the output type of the rule or transform. `when` narrowing requires type predicates, this is a typescript limitation. `(event): event is Type1 => 'field1' in event`

```ts
interface t1 {
  type: "one",
  value: string
}
interface t2 {
  type: "two",
  value: number
}
interface Event extends EventBusRuleInput<t1|t2>{}
const bus = new EventBus<Event>()

const rule1 = bus.when(this, 'rule1', (event):  event is Event<t1> => event.detail.type === "one" );

rule1.when(this, 'rule2', event => event.detail.value === "hello") // fine, value is a string
rule1.when(this, 'rule3', event => event.detail.value < 100) // fails, value in `t1` is a `string`, not a number
```

## Refining

`when` can now be chained, representing multiple AND type joins.

```ts
interface Event extends EventBusRuleInput<t1|t2>{}
const bus = new EventBus<Event>()

const rule1 = bus.when(this, 'rule1', (event):  event is Event<t1> => event.detail.type === "one" );

// same as `event.detail.type === "one" && event.detail.value === "something" && event.source === "lambda"`
const redined2 = bus
   .when(this, 'refinedRule', (event) => event.detail.value === "something")
   .when(this, 'refinedRule2', (event) => event.source === "lambda")

// both rules can be used
rule1.pipe()
redined2.pipe()
```

If a functionless event bus rule is not used, a CDK rule will not be created.

## On* Event helpers in Step Functions

https://docs.aws.amazon.com/step-functions/latest/dg/cw-events.html#cw-events-events

```ts
const machine = new StepFunction()
machine.onStarted(scope, 'started').pipe(myFunction);

// on failed and refined 
machine.onFailed(scope, 'failed').when(event => event.detail.name.startWith("SensativeEvent")).pipe(errorHandler)

// handle all negative events
machine.onStatusChanged(scope, 'allNegative')
   .when(event => event.detail.status === "FAILED" ||
       event.detail.status === "ABORTED" ||
       event.detail.status === "TIMED_OUT" )
```

BREAKING CHANGE: Event Bridge `pipe`, `when`, and `map` will now respect narrowed and declared types and may fail compilation if incorrectly configured in the past.

### CODE REVIEW

**Comment by sam-goodwin** [comment]
Should the second .when be filter? Why is this extra chain necessary?

**Comment by sam-goodwin** [comment]
Update README?

File: README.md [hunk 1]
```diff
@@ -372,6 +372,13 @@ const catPeopleEvents = bus.when(
 );
 ```
 
+Rules can be further refined by chaining multiple `when` predicates together.
+
+```ts
+// Cat people who are between 18 and 30 and do not also like dogs.
+catPeopleEvents.when(event => !event.detail.interests.includes("DOGS"))
```
[the remark this answers is not in the document]
Comment (thantos) [reply README.md:379]: `catPeopleEvents` is a rule. Maybe chaining is the wrong word?

[the remark this answers is not in the document]
Comment (sam-goodwin) [reply README.md:379]: Oh is there context above this section? I can only see this snippet

[the remark this answers is not in the document]
Comment (thantos) [reply README.md:379]: I updated the description to not mention chaining, but yes, the name was from the previous code snippet.

```ts
const catPeopleEvents = bus.when(
  (event) =>
    event["detail-type"] === "Create" &&
    event.detail.interests.includes("CATS") &&
    event.detail.age >= 18 &&
    event.detail.age < 30
);
```

[the remark this answers is not in the document]
Comment (thantos) [reply README.md:379]: new copy
> Rules can be further refined by calling `when` on a Functionless `EventBusRule`.

**Pull request merged**